src/runtime_strings.js

```diff
@@ -9,11 +9,31 @@
 // Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
 
+#if USE_PTHREADS && TEXTDECODER
+// UTF8Decoder.decode may not work with a view of a SharedArrayBuffer, see
+// https://github.com/whatwg/encoding/issues/172
+// To avoid that, we wrap around it and add a copy into a normal ArrayBuffer,
+// which can still be much faster than creating a string character by
+// character.
+function TextDecoderWrapper(encoding) {
+  var textDecoder = new TextDecoder(encoding);
+  this.decode = function(data) {
+#if ASSERTIONS
+    assert(data instanceof Uint8Array);
+#endif
+    if (data.buffer instanceof SharedArrayBuffer) {
```

Comment:
Btw, don't we know statically that this wrapper is only used on shared memory?

Reply:
Ah, this can be called on side buffers too, like filesystem buffers. Reverted the assertion and added a comment.